Add initial session-store interface and implementation

JoelSpeed opened the pull request:
## Description

This adds a SessionStore interface that can/will be used by the proxy to provide both client-side and server-side session storage implementations.

This PR is mainly copying and pasting code without much tidy up (especially in the cookie store implementation) but does add some tests to the SessionStore using ginkgo and gomega.

I haven't implemented using the `SessionStore` in the proxy itself yet so this PR should make no functional changes (bar the config changes that have been made).

I will follow with a separate PR to implement the `SessionStore` in the main proxy.

CC @brianv0

## Motivation and Context

We want to add server side sessions, see issue #138 

## How Has This Been Tested?

Test suite added for `SessionStore` to ensure cookies have correct values set and Save/Load cycle produces the same session.

I manually tested that the configuration loading changes still work, have performed full authentication flow using image built from this branch

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

syscll commented on pkg/apis/options/cookie.go:
```diff
@@ -0,0 +1,15 @@
+import "time"
+
+// CookieOptions contains configuration options relating to Cookie configuration
+type CookieOptions struct {
```
Does each field need to begin with `Cookie`? If these fields are explicitly related to a Cookie then it's a bit of unneeded duplication imo.

JoelSpeed replied:
It doesn't necessarily no, I had done this so that I didn't have to find and replace all of the occurrences of each field to minimize diffs in this PR. Also, at present this `CookieOptions` struct is embedded in the `Options` struct, if we rename these fields we probably want to get rid of this embedding, that would likely result in a change to the configuration file layout, so perhaps leave that for a later configuration oriented refactor? I have plans to replace the existing config management with viper so that we can create groups of flags and config options and compose them together for the full configuration struct

syscll replied:
Fine with me.

syscll commented on pkg/apis/sessions/interfaces.go:
```diff
@@ -0,0 +1,12 @@
+)
+
+// SessionStore is an interface to storing user sessions in the proxy
+type SessionStore interface {
```
Similar to the Cookie comment, does each method need to have the `Session` suffix as it's also a bit of unneeded duplication imo.

JoelSpeed replied:
Totally happy to rename these, these were just the first names that came off the top of my head

JoelSpeed replied:
Have renamed!

syscll commented:
This is really nice, idiomatic code :+1: 

Couple of questions:
1. Do we need to consider protecting against race conditions on the cookie store? Or will it never be called concurrently?
2. you mentioned that the cookie store is mainly copy/paste - where did it originate?
3. Do we want to test this in some sort of staging env?

JoelSpeed commented:
> Do we need to consider protecting against race conditions on the cookie store? Or will it never be called concurrently?

It definitely can be called concurrently, but at the moment we don't protect against that at all (in master that is). I'm not sure how you could lock across multiple instances to prevent it being called concurrently for a particular user so I was planning to just leave this pretty much as is. This PR plus #148 should be functionally equivalent to the existing behaviour.

I would like to add a lockable session store implementation at some point but that will likely have to be a server side session store implementation. 

The only reason this would cause a problem being called concurrently is if you are using session refreshing with the upstream provider who rotates keys (looking at OIDC here) on every refresh, then you can end up with users being logged out periodically unless you do some caching fun in front of the proxy. This has been on my mind for a while... 😅 

2. If you check #148, the actual implementation should just be all the code (pretty much) deleted from `oauth2_proxy.go` and moved into `pkg/sessions/cookie/session_store.go`

3. Im planning to build an image from #148 and will deploy it somewhere for testing, maybe internally at Pusher 🤔 I'll make sure we test it for a few days and see that everything is working as expected before we merge any of this code. I would like to (long term) set up so proper acceptance/black box tests for this project though I'm not entirely sure how to test projects with UIs, will need some research doing.

brianv0 commented on pkg/sessions/cookie/session_store.go:
```diff
@@ -0,0 +1,232 @@
+	if value != "" {
+		value = cookie.SignedValue(s.CookieSecret, s.CookieName, value, now)
+	}
+	c := s.makeCookie(req, s.CookieName, value, expiration)
```
Should the signature of `makeCookie` include a `now time.Time`?

JoelSpeed replied:
I actually made the same change in #148, do you think I should move the change across to this PR?

https://github.com/pusher/oauth2_proxy/pull/148/files#diff-6795d5c635b0b599acbb55d7272d11f8R114

brianv0 commented on pkg/sessions/cookie/session_store.go:
```diff
@@ -0,0 +1,232 @@
+// interface that stores sessions in client side cookies
+type SessionStore struct {
+	CookieCipher   *cookie.Cipher
+	CookieDomain   string
```
Should we be sending along the `options.CookieOptions` struct? This might make `makeCookie` a bit more usable across different store implementations

JoelSpeed replied:
That's a great idea! If we keep the `CookieOptions` struct we can move `makeCookie` to a shared package and make it reusable, I'll do that and join it to #148

JoelSpeed commented:
I've rebased, upped the linter timeout to stop the tests timing out, added a changelog entry and started a documentation page for the session storage stuff so will need another quick review on that last two commits before merge
